File: 03/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

const mod = 1000000007

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<20)
	writer := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer writer.Flush()

	// Читаем M и N
	line, _ := reader.ReadString('\n')
	parts := strings.Fields(strings.TrimSpace(line))
	M, _ := strconv.ParseInt(parts[0], 10, 64)
	N, _ := strconv.Atoi(parts[1])

	// Читаем потребности групп
	W := make([]int64, N)
	totalNeed := int64(0)

	// Читаем числа, обрабатывая случай, когда они могут быть в нескольких строках
	scanner := bufio.NewScanner(reader)
	scanner.Split(bufio.ScanWords)

	for i := 0; i < N && scanner.Scan(); i++ {
		val, _ := strconv.ParseInt(scanner.Text(), 10, 64)
		W[i] = val
		totalNeed += val
	}

	result := solve(M, W, totalNeed)
	writer.WriteString(fmt.Sprintf("%d\n", result))
}

// solve находит минимальную сумму квадратов недостачи
// Алгоритм: для минимизации суммы квадратов при фиксированной сумме
// нужно распределить недостачу максимально равномерно
func solve(M int64, W []int64, totalNeed int64) int {
	N := len(W)
	deficit := totalNeed - M

	if deficit == 0 {
		return 0
	}

	// Создаем пары (потребность, индекс) для сортировки
	type pair struct {
		w     int64
		index int
	}
	pairs := make([]pair, N)
	for i := 0; i < N; i++ {
		pairs[i] = pair{w: W[i], index: i}
	}

	// Сортируем по убыванию потребности
	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].w > pairs[j].w
	})

	// Распределяем недостачу оптимальным образом
	// Для минимизации суммы квадратов нужно распределить недостачу максимально равномерно
	// с учётом ограничения: недостача группы не может превышать её потребность
	origN := N
	shortfall := make([]int64, origN)
	remaining := deficit

	// Проходим по группам (отсортированным по убыванию потребности)
	// и распределяем недостачу
	// Ключевая идея: если avgShortfall <= minW среди оставшихся групп,
	// то можем распределить равномерно. Иначе, группа с минимальной потребностью
	// получает максимальную недостачу (равную своей потребности).
	for remaining > 0 && N > 0 {
		avgShortfall := remaining / int64(N)
		minW := pairs[N-1].w // минимальная потребность (pairs отсортирован по убыванию)

		// Если средняя недостача <= минимальной потребности среди оставшихся,
		// то все группы могут принять эту недостачу
		if avgShortfall <= minW {
			// Распределяем равномерно среди оставшихся групп
			baseShortfall := remaining / int64(N)
			extra := remaining % int64(N)

			// Распределяем базовую недостачу
			for j := 0; j < N; j++ {
				shortfall[pairs[j].index] = baseShortfall
			}

			// Распределяем остаток по одной единице
			for j := 0; j < int(extra); j++ {
				shortfall[pairs[j].index]++
			}
			break
		} else {
			// Группа с минимальной потребностью не может принять среднюю недостачу
			// Даем ей максимально возможную недостачу (равную потребности)
			shortfall[pairs[N-1].index] = pairs[N-1].w
			remaining -= pairs[N-1].w
			N-- // удаляем эту группу из рассмотрения
		}
	}

	N = origN // восстанавливаем для подсчёта результата

	// Вычисляем сумму квадратов недостачи по модулю
	result := int64(0)
	for i := 0; i < N; i++ {
		sq := shortfall[i] % mod
		result = (result + sq*sq%mod) % mod
	}

	return int(result)
}

File: 03/main.rs
use std::io::{self, BufRead, BufWriter, Write, Read};

const MOD: i64 = 1000000007;

fn main() {
    let stdin = io::stdin();
    let mut reader = stdin.lock();
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());

    // Читаем M и N
    let mut line = String::new();
    reader.read_line(&mut line).unwrap();
    let parts: Vec<&str> = line.trim().split_whitespace().collect();
    let m: i64 = parts[0].parse().unwrap();
    let n: usize = parts[1].parse().unwrap();

    // Читаем потребности групп
    let mut w = Vec::with_capacity(n);
    let mut total_need = 0i64;

    // Читаем числа, обрабатывая случай, когда они могут быть в нескольких строках
    let mut buffer = String::new();
    reader.read_to_string(&mut buffer).unwrap();
    let tokens: Vec<&str> = buffer.split_whitespace().collect();
    
    for i in 0..n {
        let val: i64 = tokens[i].parse().unwrap();
        w.push(val);
        total_need += val;
    }

    let result = solve(m, &w, total_need);
    writeln!(writer, "{}", result).unwrap();
    writer.flush().unwrap();
}

// solve находит минимальную сумму квадратов недостачи
// Алгоритм: для минимизации суммы квадратов при фиксированной сумме
// нужно распределить недостачу максимально равномерно
fn solve(m: i64, w: &[i64], total_need: i64) -> i64 {
    let n = w.len();
    let deficit = total_need - m;

    if deficit <= 0 {
        return 0;
    }

    // Создаем пары (потребность, индекс) для сортировки
    let mut pairs: Vec<(i64, usize)> = w.iter().enumerate().map(|(i, &w)| (w, i)).collect();

    // Сортируем по убыванию потребности
    pairs.sort_by(|a, b| b.0.cmp(&a.0));

    // Распределяем недостачу оптимальным образом
    let orig_n = n;
    let mut shortfall = vec![0i64; orig_n];
    let mut remaining = deficit;
    let mut current_n = n;

    // Проходим по группам (отсортированным по убыванию потребности)
    // и распределяем недостачу
    // Ключевая идея: если avgShortfall <= minW среди оставшихся групп,
    // то можем распределить равномерно. Иначе, группа с минимальной потребностью
    // получает максимальную недостачу (равную своей потребности).
    while remaining > 0 && current_n > 0 {
        let avg_shortfall = remaining / current_n as i64;
        let min_w = pairs[current_n - 1].0; // минимальная потребность (pairs отсортирован по убыванию)

        // Если средняя недостача <= минимальной потребности среди оставшихся,
        // то все группы могут принять эту недостачу
        if avg_shortfall <= min_w {
            // Распределяем равномерно среди оставшихся групп
            let base_shortfall = remaining / current_n as i64;
            let extra = remaining % current_n as i64;

            // Распределяем базовую недостачу
            for j in 0..current_n {
                shortfall[pairs[j].1] = base_shortfall;
            }

            // Распределяем остаток по одной единице
            for j in 0..extra as usize {
                shortfall[pairs[j].1] += 1;
            }
            break;
        } else {
            // Группа с минимальной потребностью не может принять среднюю недостачу
            // Даем ей максимально возможную недостачу (равную потребности)
            shortfall[pairs[current_n - 1].1] = pairs[current_n - 1].0;
            remaining -= pairs[current_n - 1].0;
            current_n -= 1; // удаляем эту группу из рассмотрения
        }
    }

    // Вычисляем сумму квадратов недостачи по модулю
    let mut result = 0i64;
    for &s in &shortfall {
        let sq = s % MOD;
        result = (result + sq * sq % MOD) % MOD;
    }

    result
}

